processing/src/main/java/org/apache/druid/query/operator/WindowOperatorQueryQueryToolChest.java

```diff
@@ -50,30 +57,45 @@ public class WindowOperatorQueryQueryToolChest extends QueryToolChest<RowsAndCol
   @SuppressWarnings("unchecked")
   public QueryRunner<RowsAndColumns> mergeResults(QueryRunner<RowsAndColumns> runner)
   {
-    return new RowsAndColumnsUnravelingQueryRunner(
-        (queryPlus, responseContext) -> {
-          final WindowOperatorQuery query = (WindowOperatorQuery) queryPlus.getQuery();
-          final List<OperatorFactory> opFactories = query.getOperators();
-          if (opFactories.isEmpty()) {
-            return runner.run(queryPlus, responseContext);
-          }
-
-          Supplier<Operator> opSupplier = () -> {
-            Operator retVal = new SequenceOperator(
-                runner.run(
-                    queryPlus.withQuery(query.withOperators(new ArrayList<OperatorFactory>())),
-                    responseContext
-                )
-            );
-            for (OperatorFactory operatorFactory : opFactories) {
-              retVal = operatorFactory.wrap(retVal);
-            }
-            return retVal;
-          };
+    QueryRunner<RowsAndColumns> baseRunner = (queryPlus, responseContext) -> {
+      final WindowOperatorQuery query = (WindowOperatorQuery) queryPlus.getQuery();
+      final List<OperatorFactory> opFactories = query.getOperators();
+      if (opFactories.isEmpty()) {
+        return runner.run(queryPlus, responseContext);
+      }
 
-          return new OperatorSequence(opSupplier);
+      Supplier<Operator> opSupplier = () -> {
+        Operator retVal = new SequenceOperator(
+            runner.run(
+                queryPlus.withQuery(query.withOperators(new ArrayList<OperatorFactory>())),
+                responseContext
+            )
+        );
+        for (OperatorFactory operatorFactory : opFactories) {
+          retVal = operatorFactory.wrap(retVal);
         }
-    );
+        return retVal;
+      };
+
+      return new OperatorSequence(opSupplier);
+    };
+
+    return (queryPlus, responseContext) -> {
+      final Query<RowsAndColumns> query = queryPlus.getQuery();
+      final ResultSerializationMode serializationMode = query.context().getEnum(
+          ResultSerializationMode.CTX_SERIALIZATION_PARAMETER,
+          ResultSerializationMode.class,
+          ResultSerializationMode.ROWS
+      );
+      switch (serializationMode) {
+        case FRAMES:
+          return baseRunner.run(queryPlus, responseContext);
+        case ROWS:
+          return new RowsAndColumnsUnravelingQueryRunner(baseRunner).run(queryPlus, responseContext);
```

Comment:
You are building a QueryRunner inside the `run()` method of a QueryRunner, don't do that.  Instead the QueryRunner should be able to do the right thing based on the query that it got.  In this case, you likely need to rename the `RowsAndColumnsUnravelingQueryRunner` to `RowsAndColumnsSerializationQueryRunner` and then make that query runner look at the context and do the right thing.

Additionally, when you look at this code, you will see that it's weird that, when it is asked for Frames, it returns a Sequence of RowsAndColumns.  I understand that this is because there is some other code that is going to try to handle this, but as we have learned, that method is problematic in and of itself already.  We should really move to a world where that method isn't necessary and we can do that by building the frame up here.

So, I would suggest embedding the frame-building code here and keep the other method as, effectively a noop (the identity function)